Modules/_collectionsmodule.c

```diff
@@ -1805,16 +1924,22 @@ deque_iter(dequeobject *deque)
 {
     dequeiterobject *it;
 
+    Py_BEGIN_CRITICAL_SECTION(deque);
```

Comment:
This will be simpler and avoid the `goto` if it's moved after the `if (it == NULL) {` check.

I think we can assume that `Py_TYPE(deque)` is thread-safe.

Reply:
> I think we can assume that Py_TYPE(deque) is thread-safe.

I'm not sure that's true. Instances of a subclass of `deque` can have `__class__` reassigned.

Reply:
Yes, but that's something we can handle with a stop-the-world pause when reassigning `__class__`. We don't want to bother with locks around `Py_TYPE()` accesses.